### PR TITLE
[IMP] web_editor: better tabs

### PR DESCRIPTION
**Pull request by sebgeelen**
Refactor the tabs to be :
* Automatically aligned on a 40px grid.
* A tab only need one delete to remove.
* Arrow navigation around tabs is working as intended.

+ Unit tests refactor and addition for tabs.

task-2871848

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/web_editor/views/editor.xml [hunk 1]
```diff
@@ -175,14 +175,27 @@
...
                     padding: 1px 3px;
                     font-size: 8px;
                     vertical-align: middle;
                 }
+                #mocha-report b.tab {
+                    background-color: #14386e;
+                }
+                /*
+                We need this css to make sure the tests for tabs replicate
+                the wanted behavior.
```
Comment (dmo-odoo) [inline addons/web_editor/views/editor.xml:191]: Isn't it was `base_style.scss` is about ?

File: addons/web_editor/static/src/js/editor/odoo-editor/test/editor-test.js [hunk 2]
```diff
@@ -31,36 +32,10 @@ mocha.run(failures => {
...
-        }
-
-        report.outerHTML = report.outerHTML.replaceAll('//zws//', '<b class="zws">zws</b>');
-    }
+    // Better visualisation of invisible character in test report.
+    const report = document.querySelector("#mocha-report");
+    let reportHtml = report.outerHTML
+    reportHtml = reportHtml.replaceAll('//zws//', '<b class="zws">zws</b>');
+    reportHtml = reportHtml.replaceAll('//TAB//', '<b class="tab">Tab</b>');
+    report.outerHTML = reportHtml;
```
Comment (dmo-odoo) [inline addons/web_editor/static/src/js/editor/odoo-editor/test/editor-test.js:40]: I understand tht you duplicated the logic of zws into tab but:
* Why have you removed all the rest of the code ?
* Previously this was controlled by an option. This isn't anymore. Why ?

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js [hunk 3]
```diff
@@ -3102,7 +3102,7 @@ export class OdooEditor extends EventTarget {
             // Move selection if next character is zero-width space
             if (nextCharacter === '\u200B') {
                 focusOffset += 1;
-                while (focusNode && !focusNode.textContent[focusOffset]) {
+                while (focusNode && !focusNode.textContent[focusOffset] || !closestElement(focusNode).isContentEditable) {
```
Comment (dmo-odoo) [inline addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:3105]: Why is this needed but wasn't before ?

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js [hunk 4]
```diff
@@ -3245,6 +3245,10 @@ export class OdooEditor extends EventTarget {
         for (const el of element.querySelectorAll('[contenteditable="false"]')) {
             el.setAttribute('data-oe-keep-contenteditable', '');
         }
+        // Flag elements .oe-tabs contenteditable=false.
+        for (const el of element.querySelectorAll('.oe-tabs')) {
+            el.setAttribute('contenteditable', 'false');
```
Comment (dmo-odoo) [inline addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:3250]: Why is this needed ? The only place where we create `oe-tabs` elements they already have `contentEditable=false`.

File: addons/web_editor/static/src/js/editor/odoo-editor/src/commands/deleteForward.js [hunk 5]
```diff
@@ -29,7 +31,7 @@ Text.prototype.oDeleteForward = function (offset, alreadyMoved = false) {
     if (offset === this.nodeValue.length) {
         // Delete at the end of a text node is not a specific case to
         // handle, let the element implementation handle it.
-        HTMLElement.prototype.oDeleteForward.call(this, offset, alreadyMoved);
+        HTMLElement.prototype.oDeleteForward.call(parentNode, [...parentNode.childNodes].indexOf(this) + 1);
```
Comment (dmo-odoo) [inline addons/web_editor/static/src/js/editor/odoo-editor/src/commands/deleteForward.js:34]: This is quite the change here. If there was a bug with this function then it would be better to have a specific commit with this change and a particular test that breaks with the current code but doesn't with your new one.

File: addons/web_editor/static/src/js/editor/odoo-editor/src/commands/deleteForward.js [hunk 6]
```diff
@@ -106,7 +108,15 @@ HTMLElement.prototype.oDeleteForward = function (offset) {
     }
 
     if (firstLeafNode && (isFontAwesome(firstLeafNode) || isNotEditableNode(firstLeafNode))) {
+        const nodeIsEditorTab = isEditorTab(firstLeafNode);
```
Comment (dmo-odoo) [inline addons/web_editor/static/src/js/editor/odoo-editor/src/commands/deleteForward.js:111]: Why create a variable for this ? It is only used once and the name does not add more information than the function call itself.

File: addons/web_editor/static/src/js/editor/odoo-editor/src/utils/sanitize.js [hunk 7]
```diff
@@ -184,6 +189,22 @@ class Sanitize {
                 node.textContent = '\u200B';
             }
 
+            // Ensure the editor tabs allign of the 40px grid
```
Comment (dmo-odoo) [inline addons/web_editor/static/src/js/editor/odoo-editor/src/utils/sanitize.js:192]: * allign -> align
* "align of" ? You mean "on" ? If yes then I guess it's "on *a* 40px grid" rather than *the*
* Please add a dot at the end of the sentence (and while you're at it, you can also add one to "Merge identical elements together" above, thanks ^^)

File: addons/web_editor/static/src/js/editor/odoo-editor/src/utils/sanitize.js [hunk 8]
```diff
@@ -184,6 +189,22 @@ class Sanitize {
...
+                let tabPreviousSibling = node.previousSibling;
+                while (isZWS(tabPreviousSibling)) {
+                    tabPreviousSibling = tabPreviousSibling.previousSibling;
+                }
+                if (isEditorTab(tabPreviousSibling)) {
+                    node.style.width = '40px';
+                } else {
+                    const editable = closestElement(node,'.odoo-editor-editable', true);
+                    if (editable && editable.firstElementChild) {
+                        alignNodeOnGrid(node, editable.firstElementChild);
```
Comment (dmo-odoo) [inline addons/web_editor/static/src/js/editor/odoo-editor/src/utils/sanitize.js:203]: This is only used once and is ultra specific to the particular behavior at hand here. I would inline the function here and precede it with a comment that says "// Align node on grid.". If we ever do need to align other nodes on the grid, we can make an util function then.

File: addons/web_editor/static/src/js/editor/odoo-editor/src/utils/utils.js [hunk 9]
```diff
@@ -1841,11 +1860,11 @@ export function getState(el, offset, direction, leftCType) {
     if (direction === DIRECTIONS.LEFT) {
         domPath = leftDOMPath(el, offset, reasons);
         inverseDOMPath = rightDOMPath(el, offset);
-        expr = /[^\S\u00A0]$/;
+        expr = /[^\S\u00A0\u0009]$/;
```
Comment (dmo-odoo) [inline addons/web_editor/static/src/js/editor/odoo-editor/src/utils/utils.js:1863]: I'm confused. Why do we have a `nonWhitespacesRegex` const but we write it here as well ? Because of that you had to make multiple changes where you should only have needed one. I suggest to improve this part of the code in this PR.

File: addons/web_editor/static/src/js/editor/odoo-editor/src/utils/utils.js [hunk 10]
```diff
@@ -1841,11 +1860,11 @@ export function getState(el, offset, direction, leftCType) {
...
     } else {
         domPath = rightDOMPath(el, offset, reasons);
         inverseDOMPath = leftDOMPath(el, offset);
-        expr = /^[^\S\u00A0]/;
+        expr = /^[^\S\u00A0\u0009]/;
```
Comment (dmo-odoo) [inline addons/web_editor/static/src/js/editor/odoo-editor/src/utils/utils.js:1867]: Same here.

File: addons/web_editor/static/src/js/editor/odoo-editor/src/utils/utils.js [hunk 11]
```diff
@@ -2083,10 +2102,10 @@ export function enforceWhitespace(el, offset, direction, rule) {
     let expr;
     if (direction === DIRECTIONS.LEFT) {
         domPath = leftLeafOnlyNotBlockPath(el, offset);
-        expr = /[^\S\u00A0]+$/;
+        expr = /[^\S\u00A0\u0009]+$/;
```
Comment (dmo-odoo) [inline addons/web_editor/static/src/js/editor/odoo-editor/src/utils/utils.js:2105]: Same here.

File: addons/web_editor/static/src/js/editor/odoo-editor/src/utils/utils.js [hunk 12]
```diff
@@ -2083,10 +2102,10 @@ export function enforceWhitespace(el, offset, direction, rule) {
...
     } else {
         domPath = rightLeafOnlyNotBlockPath(el, offset);
-        expr = /^[^\S\u00A0]+/;
+        expr = /^[^\S\u00A0\u0009]+/;
```
Comment (dmo-odoo) [inline addons/web_editor/static/src/js/editor/odoo-editor/src/utils/utils.js:2108]: Same here.

File: addons/web_editor/static/src/scss/web_editor.backend.scss [hunk 13]
```diff
@@ -43,6 +43,12 @@
         $oe-link-color: o-color('o-cc3-link');
         color: if($oe-link-color, $oe-link-color, #008f8c);
     }
+    .oe-tabs {
+        display: inline-block;
+        white-space: pre-wrap;
+        max-width: 40px;
+        width: 40px;
+    }
```
Comment (dmo-odoo) [inline addons/web_editor/static/src/scss/web_editor.backend.scss:51]: I think this should have been in `base_style.scss` as mentioned earlier.

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js (hunk 3 above)
Comment (sebgeelen) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:3105]: I think it's a legitimate bugfix which is not related to my change, but was not creating any visible problem prior to this change.

Basically we want to avoid sanitizing elements that are `contenteditable=false`.
I can add this change in a separe commit and/or add a comment if you want.

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js (hunk 4 above)
Comment (sebgeelen) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:3250]: We clean `contenteditable=false` during save (OdooEditor:3275) : 
```javascript 
// Remove contenteditable=false on elements
for (const el of element.querySelectorAll('[contenteditable="false"]')) {
    if (!el.hasAttribute('data-oe-keep-contenteditable')) {
        el.removeAttribute('contenteditable');
    }
}
```

So we either need to add it back when start the edition process ( current version)
Or we need to flag the `.oe-tabs` with the `data-oe-keep-contenteditable` property.

The first solution seems the best to me, because we don't really want to keep a bunch of `contenteditable=false` span out of the edition process. Seems a bit dirty IMO.

Though ?

File: addons/web_editor/static/src/js/editor/odoo-editor/src/commands/deleteForward.js (hunk 5 above)
Comment (sebgeelen) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/commands/deleteForward.js:34]: It is not realy a bug but almost, let me explain.

When we tgigger the `HTMLElement.prototype.oDeleteForward` for now we give it a node (this), so in some case the `this` in **HTMLElement** might not be a HTMLElement. which is wrong I thing.

The fact is that all the trial and error we do in those oDelete function make it not a problem because the system is fail safe enough. but I change it because I encountered a problem in the code I added if I was not able to expect an HTMLElement.

Functionally is should not change anything, and all the editor Unit tests still pass.

Obviously I can put this in his own commit.

File: addons/web_editor/static/src/js/editor/odoo-editor/src/utils/sanitize.js (hunk 7 above)
Comment (sebgeelen) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/utils/sanitize.js:192]: :grimacing:

File: addons/web_editor/static/src/js/editor/odoo-editor/src/utils/sanitize.js (hunk 8 above)
Comment (sebgeelen) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/utils/sanitize.js:203]: :nby: :grimacing:

File: addons/web_editor/static/src/js/editor/odoo-editor/src/utils/utils.js (hunk 9 above)
Comment (sebgeelen) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/utils/utils.js:1851]: They are all slightly different,  except for the `\S\u00A0\u0009` part : 
image

So I'll just add the common part in a const string and keep the inline regex if it's ok for you.

File: addons/web_editor/static/src/js/editor/odoo-editor/src/commands/deleteForward.js (hunk 5 above)
Comment (dmo-odoo) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/commands/deleteForward.js:34]: We should discuss this this afternoon.

File: addons/web_editor/static/src/js/editor/odoo-editor/src/commands/deleteBackward.js [hunk 14]
```diff
@@ -29,45 +29,60 @@ import {
...
-    const middleNode = parentNode.childNodes[firstSplitOffset];
+/**
+ * Handle text node deletion for Text.oDeleteForward and Text.oDeleteBackward.
+ *
+ * @param {int} charSize
+ * @param {int} offset
+ * @param {DIRECTIONS} direction
+ * @param {boolean} alreadyMoved
+ */
+export function textDeletion(charSize, offset, direction, alreadyMoved) {
```
Comment (dmo-odoo) [inline addons/web_editor/static/src/js/editor/odoo-editor/src/commands/deleteBackward.js:40]: `textDeletion` -> `deleteText`

Also, I'd put it in deleteForward actually, since technically "delete" is "deleteForward" as "deleteBackward" is backspace, but that's nitpicking.

File: addons/web_editor/static/src/js/editor/odoo-editor/src/commands/deleteForward.js [hunk 15]
```diff
@@ -85,7 +85,14 @@ HTMLElement.prototype.oDeleteForward = function (offset) {
     }
 
     if (firstLeafNode && (isFontAwesome(firstLeafNode) || isNotEditableNode(firstLeafNode))) {
+        const nextSibling = firstLeafNode.nextSibling;
+        const nextSiblingText = nextSibling ? nextSibling.textContent : '';
         firstLeafNode.remove();
+        if (isEditorTab(firstLeafNode) && nextSiblingText[0] === '\u200B') {
+            // When deleting an editor tab,
+            // we need to ensure it's related ZWS il deleted as well.
```
Comment (dmo-odoo) [inline addons/web_editor/static/src/js/editor/odoo-editor/src/commands/deleteForward.js:93]: wrap

File: addons/web_editor/static/src/js/editor/odoo-editor/src/utils/sanitize.js [hunk 16]
```diff
@@ -107,10 +110,11 @@ class Sanitize {
 
     _parse(node) {
         while (node) {
-            // Merge identical elements together
+            // Merge identical elements together.
```
Comment (dmo-odoo) [inline addons/web_editor/static/src/js/editor/odoo-editor/src/utils/sanitize.js:113]: ❤️

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js [hunk 17]
```diff
@@ -3031,7 +3031,7 @@ export class OdooEditor extends EventTarget {
             } else if (closestTag === 'TABLE') {
                 this._onTabulationInTable(ev);
             } else if (!ev.shiftKey) {
-                this.execCommand('insert', '\u00A0 \u00A0\u00A0');
+                 this.execCommand('insert', parseHTML('<span class="oe-tabs" contenteditable="false">\u0009</span>\u200B'));
```
Comment (dmo-odoo) [inline addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:3034]: This line looks weirdly indented on github. Is it a display glitch or is the indentation actually incorrect ?

Comment (sebgeelen) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:3034]: indeed, must be a rebase conflict resolution "error" again. sorry

**Comment by dmo-odoo** [comment]
@robodoo rebase-ff r+
